### DISCUSSION TITLE
Creating a uniform and using it in the shader

### DISCUSSION THREAD

**Discussion by dov** [vsg-dev/VulkanSceneGraph#1243] [General]
In my quest of learning VSG I have extracted from the example source code that to create a uniform that I can change every frame I need to do the following four steps in C++:

1. Declare struct `ToyUniform` containing the uniform variables
```
struct ToyUniform {
    vsg::ivec2 iResolution;
}
```
2. Wrap the scruct in a `using ToyUniformValue = vsg::Value<ToyUniform>`
3. Allocate by `create` and assign set it up as follows:
```
  toyUniform = ToyUniformValue::create();
  toyUniform->properties.dataVariance = vsg::DataVariance::DYNAMIC_DATA;
  toyUniform->value().iResolution = {800,600};
```
4. Create a uniform descriptor:
```
  auto toyUniformDescriptor = vsg::DescriptorBuffer::create(toyUniform, 0, 0, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER);
```
5. Add the uniform descriptor to the descriptor set:
```
  auto descriptorSet = vsg::DescriptorSet::create(descriptorSetLayout, vsg::Descriptors{toyUniformDescriptor});
```
The `descriptorSet` is then used as in `vsgdraw.cpp`.

On the shader side I declared it in a uniform and used it as follows in the vertex shader:

```
layout(location = 0) in vec2 inVertex;
layout(location = 1) in vec2 inTexture;

layout(set = 0, binding = 0) uniform UBO {
    ivec2 iResolution;
} ubo;
:
void main()
{
    gl_Position = ...; // This is ok since my fragment shader paints in the entire quad!
}

```
However, in the fragment shader, by writing the following shader I realized that the uniform is not initialized and the following shader outputs a constant red image. 

```
#version 450

layout(set = 0, binding = 0) uniform UBO {
    ivec2 iResolution;
} ubo;

layout(location = 0) in vec2 fragCoord;
layout(location = 0) out vec4 fragColor;

void main()
{
  if (abs(ubo.iResolution.x) == 0)
      fragColor = vec4(1.0,0.0,0.0,1.0); // red
  else
      fragColor = vec4(0.0,1.0,0.0,1.0); // green
}
```

What am I missing?

Thanks in advance!

**Comment by robertosfield** [comment]
Broadly it looks OK, but we don't know what else is being doing by the rest of your application, perhaps it has another DescrptorSet that is being applied on set=0, binding=0.

**Reply by dov** [reply]
Thanks for confirming that I'm on the right track!

I uploaded the source code and the two shaders to the following gist:

https://gist.github.com/dov/f7d8e0adf755d6ced43ce685efeb1c43

If you have some more feedback on my code, I'd be happy to receive it!

**Reply by dov** [reply]
The mystery (in my mind at least!) deepens. I tried inspecting my application with renderdoc and it seems that the value of `iResolution` is correctly initialized in the fragment shader. So why does my if shader branch according to it being zero below and paint the output red?

```
#version 450

layout(set = 0, binding = 0) uniform UBO {
    ivec2 iResolution;
    vec2 iMouse;
    float iTime;
} ubo;

layout(location = 0) in vec2 fragCoord;
layout(location = 0) out vec4 fragColor;

void main()
{
  // Do a dummy test to see if iResolution has been initialized
  if (ubo.iResolution.x == 0)
      fragColor = vec4(1.0,0.0,0.0,1.0); // red
  else
      fragColor = vec4(0.0,1.0,0.0,1.0); // green
}

```
![uniform](https://github.com/user-attachments/assets/8d8c3c1c-4c8c-48a7-97ce-e1b4f44c3ec2)

**Reply by dov** [reply]
I found it! My `descriptorBindings` was wrong! The correct contents should be:
```
  vsg::DescriptorSetLayoutBindings descriptorBindings{
    {0, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER, 1, 
    VK_SHADER_STAGE_VERTEX_BIT|VK_SHADER_STAGE_FRAGMENT_BIT, nullptr} // { binding, descriptorType, descriptorCount, stageFlags, pImmutableSamplers}
  };

  auto descriptorSetLayout = vsg::DescriptorSetLayout::create(descriptorBindings);
```
with this in place my uniforms and shaders behave as expected. :-)

This should be added to my original uniform "recipe".

**Reply by robertosfield** [reply]
Great to hear you've spotted the problem.

Tracking down this type of issue is often helped by using the Vulkan Validation Layer.  You can enable this by using the -d command line option in many of the examples like vsgviewer, or programatically via setting windowTraits::debugLayer = true;

**Comment by dov** [comment]
I finished my vsgshadertoy example, which allows rendering a shadertoy.com like shader in a vsg window.

I submitted the result as a pull request to vsgExamples. I'd be happy to receive feedback about how I can improve my VSG usage and understanding.
